src/indicators.pyx

```diff
@@ -92,17 +94,88 @@ cpdef float WMA(double[:] closes, int period, float weightedFactor):
         return ma
 
 
+cpdef DERIVATIVE(double[:] values, int period): #TODO additional testing 
+        """
+        Derivate function... interpolates a list of values, derives the list, and returns d/dx list 
+        @param values: to interpolate & derive
+        @param period: period to calculate for
+        """
+        cdef int length = values.shape[0]
+        cdef int j = 0
+        cdef double[:] xvals = np.zeros((length-period,))
+        cdef double[:] yvals = np.zeros((length-period,))
+        for i in xrange(length - period, length):
+                xvals[j] = j
+                yvals[j] = SMA(values[0 : i], period)
+                j+=1
+        dx = np.diff(xvals)
+        dy = np.diff(yvals)
 
-cpdef FIB_BANDS(double[:] closes, int bb_period):
+        return (dy/dx)
+        
+cpdef list EMA(double[:] closes, int period, float alpha = .3, int epsilon = 0):
+        """
+        Exponential moving average function 
+        @param closes: list of closing candle prices 
+        @param period: period to calculate for 
+        @param alpha: alpha value
+        @param epsilon: epsilon value 
+        """
+
+        assert(0 < alpha < 1, ("out of range, alpha='%s'" % alpha))
+
+        assert(not 0 <= epsilon < alpha, ("out of range, epsilon='%s'" % epsilon))
+
+        assert(period>0, ("out of range, period='%s'" %period))
+
+        cdef float currentWeight
+        cdef float numerator      
+        cdef float denominator 
+        cdef int i 
+        cdef int j 
+        cdef int k = 0
+        cdef int length = closes.shape[0]
+        cdef double *results = <double *> malloc(sizeof(double)* (period+1))
+
+        for i in xrange(length-period, length):
+                currentWeight = 1.0
+                numerator = 0.0 
+                denominator = 0.0
+                for j in xrange(i, 0, -1):
+                        numerator += closes[j] * currentWeight
+                        denominator += currentWeight
+
+                        currentWeight *= alpha
+                        if currentWeight < epsilon:
+                                break
+                results[k] = numerator / denominator
+                k+=1
```

Comment:
You can do `for in` loops in cython 👀
If it works the same as regular python you shouldn't need to declare `cdef int i/j`. Unless that explicitly types the values i and j hold i guess

Reply:
i has to be an int for this instance since j iterates from i to 0